pandas/core/frame.py

```diff
@@ -5034,6 +5034,51 @@ def drop(  # type: ignore[override]
             errors=errors,
         )
 
+    @overload
+    def rename(
+        self,
+        mapper: Renamer | None = ...,
+        *,
+        index: Renamer | None = ...,
+        columns: Renamer | None = ...,
```

Comment:
See https://github.com/python/mypy/issues/12464
If someone writes:
```python
df = pd.DataFrame(columns=["a"])
col_map = {"a": "b"}
df.rename(columns=col_map)
```
mypy reports an error.  So our typing here isn't going to work for this common use case.

Reply:
May have to change `Renamer` to use `Mapping[Any, Any]` because `Mapping` is invariant.

Reply:
Another solution is to leave `Renamer` alone, and force users to type their dict:
```python
df = pd.DataFrame(columns=["a"])
col_map: Dict[Hashable, str] = {"a": "b"}
df.rename(columns=col_map)
```

Reply:
Sorry for the force push. I implemented an (ugly) TypeVar magic and your tests seems to work with both mypy and pyright :)

```py
# pyright: strict
import pandas as pd

def test_types_rename() -> None:
    df = pd.DataFrame(columns=["a"])
    col_map = {"a": "b"}
    df.rename(columns=col_map)
    df.rename(columns={"a": "b"})
    df.rename(columns={1: "b"})
    # Apparently all of these calls are accepted by pandas
    df.rename(columns={None: "b"})
    df.rename(columns={object(): "b"})
    df.rename(columns={(2, 1): "b"})
```

Reply:
Sorry for the force push again (using `git merge upstream/main` multiple times somehow messed things up on my side). I'll mark this as resolved to be able to keep an overview of this PR.